src/prefect/events/clients.py

```diff
@@ -0,0 +1,191 @@
+import abc
+import asyncio
+from types import TracebackType
+from typing import Any, ClassVar, Dict, List, Optional, Tuple, Type
+
+from websockets.client import WebSocketClientProtocol, connect
+from websockets.exceptions import ConnectionClosed
+
+from prefect.events import Event
+
+
+class EventsClient(abc.ABC):
+    """The abstract interface for all Prefect Events clients"""
+
+    @abc.abstractmethod
+    async def emit(self, event: Event) -> None:  # pragma: no cover
+        """Emit a single event"""
+
+    def _require_context(self) -> None:
+        if not hasattr(self, "_in_context"):
+            raise TypeError(
+                "Events may only be emitted while this client is being used as a "
+                "context manager"
+            )
+
+    async def __aenter__(self) -> "EventsClient":
+        self._in_context = True
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: Optional[Type[Exception]],
+        exc_val: Optional[Exception],
+        exc_tb: Optional[TracebackType],
+    ) -> None:
+        del self._in_context
+        return None
+
+
+class NullEventsClient(EventsClient):
+    """A Prefect Events client implementation that does nothing"""
+
+    async def emit(self, event: Event) -> None:
+        """Emit a single event"""
+        self._require_context()
```

Comment:
perhaps it's worth standardizing this handling somewhere? either a decorator or allowing a separate `_emit` method for implementation?

Reply:
Yeah this thing got a little ugly, I like what you're saying about having `emit` be concrete with an abstract `_emit`.  That would be much cleaner